### PR TITLE
doc/nat-techniques: detail out the results of research

### PR DESCRIPTION
**Pull request by ustulation**
Explain how various techniques were implemented or innovated for punching through NAT.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1,3 +1,187 @@
...
+//!
+//! This is the least restrictive NAT. All we need to do is make sure the mapping exists. This can
+//! be done by simply going through the rendezvous handshake coded in this crate. It should have
+//! the same effect as if the port had been forwarded manually or via [Internet Gateway Device
+//! Protocol][0] through code. If both sides are full cone then we can easily establish p2p
+//! communication.
+//!
+//! ### Endpoint Address & Port Dependent Filtering (Port Restricted Cone)
+//!
+//! This is the most restrictive NAT under `EIM`. In this crate this is what we assume for `EIMs`
```
Comment (povilasb) [inline src/lib.rs]: Ticks should be used for code. So instead of  ``EIM ` maybe we could use * EIM *?

File: src/lib.rs [hunk 2]
```diff
@@ -1,3 +1,187 @@
+//! # General
+//!
+//! The goal of this crate is to provide a robust and crypto-secure NAT traversal for peer to peer
+//! connection. It assumes publicly reachable rendezvous servers are provided. The server code
+//! itself is in the crate too, so the crate can either be used to deploy a server or used for peer
+//! to peer client communication or both simultaneously - for e.g. if you run the server on a port
+//! forwarded endpoint, it will be publicly available for others to rendezvous while you could
+//! choose normal NAT traversal mechanisms to communicate with other peers.
+//!
+//! ## Endpoint Independent Mapping (EIM)
```
Comment (povilasb) [inline src/lib.rs]: It would be nice to have a couple of sentences mentioning there are different NAT types before going deeper into that.

Comment (ustulation) [reply src/lib.rs]: But EIM etc _are_ the diff NAT types @povilasb .. check around page-6 [here](https://tools.ietf.org/html/rfc4787)

Comment (ustulation) [reply src/lib.rs]: Ah never mind - i think you meant just add a line explaining the upcoming sections are the various types of NAT of something. I did that, pls see if OK.

**Review by povilasb — APPROVED** [review]

**Pull request merged**